### ISSUE TITLE
JStree problems create_node

### ISSUE THREAD

**Issue by wrustel** [vakata/jstree#983]
Hi, Im doing an update dynamically with ajax, I get right the data from the server and I call the method create_node to make the childrens from the parent but this function never return true, never make a node this is what I do:

 $.ajax({
			dataType: "json",
		            url : 'http://localhost:8080/test-web/space/newnodes/'+data.node.id,
		            type: "GET",
		            success : function(ops) {		            	 
		                  data = []
		                  for( opnum in ops ){
		                    var op = ops[opnum]
		                    node = {
		                        "data" : op.info,
		                        "metadata" :  op ,
		                        "state" : "closed"		                        
		                    }	     
		                    var parent = $('#selector').jstree('get_selected');
		                    var newNode = { id: op.id,parent : op.parent, text: op.text, state: "open", data: op.info };
		                    $("#treeDiv").jstree(true).create_node("300", newNode);
		                    var CurrentNode = $("#selector").jstree("get_selected");
		                    var id = $("#treeDiv").jstree('create_node', op.parent, newNode, 'last',null,null);
		                    var id2=  $('#treeDiv').jstree().create_node(op.parent , JSON.stringify(op),'last' );
		                   
		                  }
		                  $('#treeDiv').jstree(true).open_all();
		                  return data;     
		            }
		        })

can you help me?. I did call_back true in the constructor of my jstree, I tested everything that I read in internet and I cant draw the childrens, I debugged in chrome and I get correctly the information from the server but this function does not work anyway. Also I tested a function call with json format or object.

**Comment by vakata** [comment]
Make sure `check_callback` is set to `true`, if so - all should work properly unless you have some `types` rules which prevent the creation of the node.
Here is a demo: http://jsfiddle.net/DGAF4/272/

**Comment by wrustel** [comment]
$('#treeDiv')
					.jstree({
						"core" : {
							"animation" : 0,
							"check_callback" : true,	
							"themes" : { "stripes" : true },
						},
						 "checkbox": {
					            "keep_selected_style": true
					        },
						"plugins" : [ "contextmenu","wholerow","types","dnd","state" ]					     
					}).jstree("open_all");

this is how I have defined the jstree div

**Comment by vakata** [comment]
Hi, once again - here is a demo with your code: http://jsfiddle.net/DGAF4/273/
Can you provide a demo which has the problem - I really can't reproduce. As said before, maybe you have a types config which prevents the creation of the node?

**Comment by wrustel** [comment]
When I create node with right click -> create node, there isnt any problem, all works fine. But I dont want to do that, I want to click in a node get via ajax the server response with news nodes information and draw the childrens, it's just a virtual directory stored in a data base.

Here is my whole code:


div container:



```html
<div id="treediv">					
		<ul id= "#">
		<li id = "151"><a href = "#" >father</a>
		<ul>
		<li id = "300"	><a href = "#" >children</li>
		 <ul></ul>
		</ul>	
		</li>				    
		</ul>					 
	</div>
```



Scripts: 
var updateTree = function(data,cb){		
				
				$.ajax({
				  dataType: "json",
		            url : 'http://localhost:8080/arquivo-web/espaciofisico/despliegue/'+data.node.id,
		            type: "GET",
		            success : function(ops) {	
		            	 data = [];
		                  for( opnum in ops ){
		                    var op = ops[opnum]
		                    node = {
		                        "data" : op.info,
		                        "metadata" :  op ,
		                        "state" : "closed"		                        
		                    }		     
		                    var result = $('#treediv').jstree(true).create_node( {'id':op.parent},node.metadata,"inside");
		                    data.push( node );
		                  }       
		            	 
		                  $('#treediv').jstree(true).open_all();
		                  $('#treediv').jstree(true).refresh();
		                  return data;  
		            }
		            })	             
		      }//END
				
			 $(function () { $('#treediv').jstree(); });				 
				
					$('#treediv')
					.jstree({
						"core" : {
							"animation" : 0,
							'check_callback' : "true",	
							'themes' : { "stripes" : true },
						},
						"plugins" : [ "contextmenu","wholerow","types","dnd","state" ]					     
					}).jstree("open_all");
					
					 $('#treediv').on('select_node.jstree',function (e,data){
						 setTimeout(updateTree(data,cb),300);
					  });
					 
					var cb =  function(){
						$('#treediv').jstree("refresh");
					}

**Comment by vakata** [comment]
Did you read the docs on the create node function: http://www.jstree.com/api/#/?q=create_node&f=create_node%28%5Bobj,%20node,%20pos,%20callback,%20is_loaded%5D%29
1) "inside" is not a valid position for `create_node`
2) The first parameter is simply the parent ID (or a DOM object), not a plain object (altough it MIGHT work)
3) Make sure the parent parameter actually exists in the tree
4) check_callback should be boolean true, not string "true"

Also from your code there is absolutely no way to determine if the above condition 3) is met. Please make corrections according to each of the 4 points and see if it works - I guarantee that if all 4 points are corrected everything will work fine.

Best regards,
Ivan

**Comment by wrustel** [comment]
I changed my code and now there is a little bit difference.

At first I load the tree from AJAX, this works fine:

this is the result from ajax call:

 "[{"id":"151","parent":"#","text":"sssd","icon":null,"state":{"opened":true,"disabled":false,"selected":false},"childrens":null},{"id":"300","parent":"151","text":"hijo","icon":null,"state":null,"childrens":null}]"

```javascript
$(function () { 	$('#treediv')
					.jstree({
						"core" : {
							'data' : {
							    'url' : 'http://localhost:8080/arquivo-web/espaciofisico/despliegue/'+0,
							    'data' : function (nodo) {
							    	var res;
							    		$.ajax({
										  dataType: "json",
								            url : 'http://localhost:8080/arquivo-web/espaciofisico/despliegue/'+0,
								            type: "GET",
								            success : function(ops) {	
								            	 var res = JSON.stringify(ops);      
								            }})		
								         return res;
							    },
							  },
							"animation" : 0,
							'check_callback' : true,	
							'themes' : { "stripes" : true },
						},						
						"plugins" : [ "contextmenu","wholerow","types","dnd","state" ]	,
						
					});  });
```

``` javascript

then I do subscribe to the event select_node
 $('#treediv').on('select_node.jstree',function (e,data){
		 setTimeout(updateTree(data,cb),300); 
	});
```

and in the end this is what should to do the function updateTree 

``` javascript
 var updateTree = function(data,cb){						
				$.ajax({
				  dataType: "json",
		            url : 'http://localhost:8080/arquivo-web/espaciofisico/despliegue/'+data.node.id,
		            type: "GET",
		            success : function(ops) {	
		            	 data = [];
		            	 var consulted = 0;
		                  for( opnum in ops ){
		                    var op = ops[opnum]
		                    node = {
		                        "data" : op.info,
		                        "metadata" :  op ,
		                        "state" : "closed"		                        
		                    }		     
		                    var result = $('#treediv').jstree(true).create_node( {'id':op.parent},node.metadata,"last");
		                    data.push( node );
		                    consulted = op.parent;
		                  }       
		                  consultedId[size] = consulted;
		                  size++;
		                  return data;  
		            }
		            })	             
		      }//END

var cb =  function(){
	$('#treediv').jstree("refresh");
}
```

Value of op.parent: "300"
Value of node.metadata: 

childrens: null
icon: null
id: "350"
parent: "300"
state: Object
disabled: false
opened: true
selected: false
text: "nieto"
__proto__: Object

And this is when doesnt work the function create_node, I means the function return the new children ID, but doesnt draw anything.

Kind regards

**Comment by vakata** [comment]
The `core.data.data` setting is supposed to provide parameters for the request, not make a request itself. The way you defined it - it makes absolutely no sense and has no effect even if removed. 
I will try to reproduce your issue and let you know, please provide a sample of what: `'http://localhost:8080/arquivo-web/espaciofisico/despliegue/'+data.node.id` 
would return.

Also keep in mind - it seems to me you are trying to do lazy loading manually, and in a wrong way at that. Did you ahve a look at the builtin lazy loading features?

**Comment by wrustel** [comment]
The url return[{"id":"151","parent":"#","text":"sssd","icon":null,"state":{"opened":true,"disabled":false,"selected":false},"childrens":null},{"id":"300","parent":"151","text":"hijo","icon":null,"state":null,"childrens":null}]" this works well I means, this code draw the tree well. But when I click in one of the nodes, for example the node with ID 300, in the method create_node() the function get_node(300) never find the node made by lazy way.

**Comment by vakata** [comment]
Sorry, this is way too messy. Provided the parent is there - `create_node` will work. 

At this point I can only help you if I see a live demo of this, I do not have the time to provide more jsfiddles. 

I will spare some of my time to help you debug your code - that is not a problem, but I am certain the problem is not with `create_node`. So if you need my help debugging your code, please provide a live demo I will see what I can do.

Best regards,
Ivan

**Comment by wrustel** [comment]
Sorry for the delay, here is a live example mocking the result from the server

http://jsfiddle.net/DGAF4/287/

**Comment by vakata** [comment]
In the demo you provided the button works just fine - where is the problem? (also keep in mind there is no `"childrens"` key, especially when using the alternative JSON format).
Anyway, so provided this is a demo of the problem - how do I reproduce the issue. Right now when I click "new" I can see the new node being created and its ID is logged in the console.

**Comment by wrustel** [comment]
Instead to make the childrens manually with a click on button I want to do it from an AJAX call.

**Comment by vakata** [comment]
OK, let me clarify - in order to help you debug your own code, I need to see a demo of the actual problem. If you need my assistance - please provide a demo.

**Comment by wrustel** [comment]
Sorry for the delay,

I fixed by myself the issue. The main problem was that using spring mvc 4 framework there are a conflict between jquery and plugins if you dont define the import to jquery correctly.


I was getting and unusual error "Undefined function"  from jquery, ajax and create_node. But finally after thinking about how I should import and use currectly jquery I found the solution.

Thank you for your time.

By the way, where can I find the different kind of icon's name to change them (for example, "file" draw files)?

kind regards
